File: 039. Combination Sum/combinationSum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    var comb = (can,tar,res,res_list) => {
        if (tar==0) {
            res_list.push([...res])
            return
        }
        else if (tar>0) {
            // [...res, can[i]] not res+[can[i]]
            // res+[can[i]] : [3,12] -> [3,1,2]
            for (let i=0; i<can.length; i++) {
                comb(can.slice(i), tar-can[i], res+[can[i]], res_list)
            }
        }
        else {
            return
        }
    }

    res_list = []
    comb(candidates,target,[],res_list)
    return res_list
};

// Runtime 145 ms / Memory 48.5 MB
// debug: https://jsfiddle.net/

File: 039. Combination Sum/combinationSum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res_list;
        vector<int> res;
        comb(0,candidates,target,res,res_list);
        return res_list;
    }
    void comb(int idx, vector<int> can, int tar, vector<int> &res, vector<vector<int>> &res_list) {
        // match: output
        if (tar==0) { res_list.push_back(res); }
        // no match: overflow
        else if(tar<0) { return; }
        // no match: fill up
        for(int i=idx; i<can.size(); i++) {
            res.push_back(can[i]); //nums array slice
            comb(i,can,tar-can[i],res,res_list);
            res.pop_back();
        }
    }
};